# Ejercicio QKD
Cifrado de Información (CC3078) - Sección 10  
Profesor Ludwing Cano  
Erick Stiv Junior Guerra Muñoz - 21781  

29 de mayo de 2025

### Importación de librerías necesarias

In [3]:
import pandas as pd
from random import randint

In [7]:
def codificar_foton(bit, base):
    if base == '+':
        return '↕' if bit == 0 else '↔'
    elif base == 'x':
        return '↖' if bit == 0 else '↘'

In [20]:
bases = ['+', 'x']

data = [{
    'Índice': 1,
    'Bit de Alice': 1,
    'Base de Alice': bases[1],
    'Fotón enviado': codificar_foton(1, bases[1]),
    'Base de Bob': bases[0],
    'Bit generado': 1,
    '¿Bases coinciden?': bases[1] == bases[0],
    '¿Usar bit?': bases[1] == bases[0]
  },{
    'Índice': 2,
    'Bit de Alice': 1,
    'Base de Alice': bases[1],
    'Fotón enviado': codificar_foton(1, bases[1]),
    'Base de Bob': bases[1],
    'Bit generado': 1,
    '¿Bases coinciden?': bases[1] == bases[1],
    '¿Usar bit?': bases[1] == bases[1]
  }]
for i in range(2, 15):
  bit_alice = randint(0, 1)
  base_alice = bases[randint(0, 1)]

  foton_enviado = codificar_foton(bit_alice, base_alice)
  base_bob = bases[randint(0, 1)]

  coinciden = base_alice == base_bob
  usar_bit = base_alice == base_bob
  
  bit_bob = bit_alice if base_alice == base_bob else randint(0, 1)

  data.append({
    'Índice': i+1,
    'Bit de Alice': bit_alice,
    'Base de Alice': base_alice,
    'Fotón enviado': foton_enviado,
    'Base de Bob': base_bob,
    'Bit generado': bit_bob,
    '¿Bases coinciden?': coinciden,
    '¿Usar bit?': usar_bit
  })

df = pd.DataFrame(data)

In [23]:
df

,Índice,Bit de Alice,Base de Alice,Fotón enviado,Base de Bob,Bit generado,¿Bases coinciden?,¿Usar bit?
0,1,1,x,↘,+,1,False,False
1,2,1,x,↘,x,1,True,True
2,3,1,+,↔,x,1,False,False
3,4,0,+,↕,x,0,False,False
4,5,1,x,↘,x,1,True,True
5,6,0,+,↕,+,0,True,True
6,7,1,x,↘,+,1,False,False
7,8,1,x,↘,+,0,False,False
8,9,0,+,↕,+,0,True,True
9,10,1,x,↘,x,1,True,True


In [24]:
key_bits = df[df['¿Usar bit?']]['Bit generado'].count()
print("Bits finales de clave:", key_bits)
print("Porcentaje del total:", (key_bits / len(df)) * 100, "%")

Bits finales de clave: 8
Porcentaje del total: 53.333333333333336 %


### Análisis

* ¿Cuántos bits finales obtuvieron de la clave?  
**R:** Se obtuvieron tan solo 8 bits en total.

* ¿Qué porcentaje representa respecto al total?  
**R:** Representa un aproximado del 53.33% del conjunto total de  registros generados.

* ¿Qué pasaría si Eve interceptara los fotones y usara bases incorrectas?  
**R:** Si Eve interceptara los fotones, se enfrentaría a un problema similar al de Bob puesto que tampoco conocería cuál fue la base utilizada por Alice. En este caso, al intentar adivinar la base, tendría un 50% de probabilidades de interpretar el fotón de manera incorrecta, por lo que, al reenviar el fotón a Bob, este podría detectar la presencia de una "espía" al verificar con Alice que las bases son iguales pero el bit es distinto.

* ¿Cómo se puede detectar su presencia?  
**R:** Como se mencionó anteriormente, cuando Alice y Bob comparen sus bases y observen que, a pesar de ser iguales, los bits difieren, esto evidenciaría la presencia de un tercero interceptando sus mensajes.

* ¿Qué ventajas y desventajas tiene este protocolo frente a otros cifrados tradicionales?  
**R:** Como ventajas, se destaca el uso de principios físicos para determinar el valor de cada bit en la clave, puesto que esto agrega una mayor aleatoriedad al proceso, evitando que alguien intercepte la clave de forma correcta. Además, presenta invulnerabilidades para ataques incluso originados en computadoras cuánticas, ya que cualquier anomalía causada por un interceptor sería fácilmente detectable y medible para tomar las medidas necesarias.    

    Por otro lado, como desventajas podemos mencionar la complejidad de su infraestructura e implementación, ya que se requiere hardware especializado para sistemas cuánticos, por lo que requiere de mayores costos y no es fácilmente aplicable en equipos convencionales. Por otro lado, al ser un proceso extenso, actualmente funciona tan solo en distancias limitadas mediante fibra óptica, causando también que la transmisión de claves completas sea mucho más lenta que en métodos tradicionales.